redskyctl/internal/commands/experiments/experiments.go

```diff
@@ -101,47 +105,100 @@ func (n *name) experimentName() experimentsv1alpha1.ExperimentName {
+// number is present.
+func (n *name) trialNumber() int64 {
+	return n.Number
+}
 
 // parseNames parses a list of arguments into structured names
 func parseNames(args []string) ([]name, error) {
-	var t resourceType
 	names := make([]name, 0, len(args))
+
+	var defaultType string
 	for _, arg := range args {
-		n := name{Type: t, Name: arg, Number: -1}
-		if sm := numberSuffixPattern.FindStringSubmatch(n.Name); sm != nil && sm[2] != "" {
-			n.Number, _ = strconv.ParseInt(sm[2], 10, 64)
-			n.Name = sm[1]
+		// Split the argument into type/name/number
+		argType, argName, argNumber := splitArg(arg, defaultType)
+		argIsTypeOnly := argName == "" && argNumber == ""
+
+		// Normalize the type
+		normalType, pluralType, err := normalizeType(argType)
+		if err != nil {
+			return nil, err
+		}
+
+		// Set the default type if it hasn't been set and no name was supplied
+		if defaultType == "" && argIsTypeOnly {
+			defaultType = pluralType
+			continue
+		}
+
+		// Create a new name
+		n := name{Type: normalType, Name: argName, Number: -1}
+
+		// Special case where trial can alternatively end with "-<NUM>" instead of "/<NUM>"
+		if n.Type == typeTrial && argNumber == "" && pluralType == string(normalType) {
+			if pos := strings.LastIndex(argName, "-"); pos > 0 {
+				n.Name, argNumber = argName[0:pos], argName[pos+1:]
+			}
 		}
 
-		p := strings.SplitN(n.Name, "/", 2)
-		if len(p) > 1 || t == "" {
-			nt, err := normalizeType(p[0])
+		// Parse the number, if present
+		if argNumber != "" {
+			if n.Type != typeTrial {
+				return nil, fmt.Errorf("%s name cannot include a number: %s", n.Type, arg)
+			}
+			n.Number, err = strconv.ParseInt(argNumber, 10, 64)
 			if err != nil {
 				return nil, err
 			}
-			if len(p) > 1 {
-				n.Type = nt
-				n.Name = p[1]
-			} else if t == "" {
-				t = nt
-				continue
-			}
 		}
+
 		names = append(names, n)
 	}
 
+	// If no names were generated we can just use the default type
 	if len(names) == 0 {
-		if t == "" {
+		if defaultType == "" {
 			return nil, fmt.Errorf("required resource not specified")
 		}
-		names = append(names, name{Type: t, Number: -1})
+		normalType, _, err := normalizeType(defaultType)
+		if err != nil {
+			return nil, err
+		}
+		return []name{{Type: normalType, Number: -1}}, nil
 	}
 
 	return names, nil
 }
 
+func splitArg(arg, defType string) (parsedType string, parsedName string, parsedNumber string) {
+	p := strings.Split(arg, "/")
+	switch len(p) {
+	case 1:
+		// type | name
+		if defType != "" {
+			return defType, p[0], ""
+		}
+		return p[0], "", ""
+	case 2:
+		// type/name | name/number
+		if _, err := strconv.ParseInt(p[1], 10, 64); err != nil {
+			return p[0], p[1], ""
+		}
+		return defType, p[0], p[1]
+	case 3:
+		// type/name/number
+		if p[2] == "" {
+			p[2] = "-1"
+		}
+		return p[0], p[1], p[2]
+	default:
+		// likely invalid name
+		return p[0], strings.Join(p[1:], "/"), ""
```

Comment:
Are there any cases off the top of your head where this would be ok?

Should we just fail here instead?